sys/net/routing/ng_rpl/ng_rpl.c

```diff
@@ -0,0 +1,314 @@
+    while (1) {
+        DEBUG("RPL: waiting for incoming message.\n");
+        msg_receive(&msg);
+
+        switch (msg.type) {
+            case NG_RPL_MSG_TYPE_LIFETIME_UPDATE:
+                DEBUG("RPL: NG_RPL_MSG_TYPE_LIFETIME_UPDATE received\n");
+                _update_lifetime();
+                break;
+            case NG_RPL_MSG_TYPE_TRICKLE_INTERVAL:
+                DEBUG("RPL: NG_RPL_MSG_TYPE_TRICKLE_INTERVAL received\n");
+                trickle = (trickle_t *) msg.content.ptr;
+                if (trickle && (trickle->callback.func != NULL)) {
+                    trickle_interval(trickle);
+                }
+                break;
+            case NG_RPL_MSG_TYPE_TRICKLE_CALLBACK:
+                DEBUG("RPL: NG_RPL_MSG_TYPE_TRICKLE_CALLBACK received\n");
+                trickle = (trickle_t *) msg.content.ptr;
+                if (trickle && (trickle->callback.func != NULL)) {
+                    trickle_callback(trickle);
+                }
+                break;
+            case NG_RPL_MSG_TYPE_DAO_HANDLE:
+                DEBUG("RPL: NG_RPL_MSG_TYPE_DAO_HANDLE received\n");
+                dodag = (ng_rpl_dodag_t *) msg.content.ptr;
+                if (dodag && (dodag->state != 0)) {
+                    _dao_handle_send(dodag);
+                }
+                break;
+            case NG_RPL_MSG_TYPE_CLEANUP_HANDLE:
+                DEBUG("RPL: NG_RPL_MSG_TYPE_CLEANUP received\n");
+                dodag = (ng_rpl_dodag_t *) msg.content.ptr;
+                if (dodag && (dodag->state != 0) && (dodag->parents == NULL)) {
+                    /* no parents - delete this DODAG */
+                    ng_rpl_dodag_remove(dodag);
+                }
+                break;
+            case NG_NETAPI_MSG_TYPE_RCV:
+                DEBUG("RPL: NG_NETAPI_MSG_TYPE_RCV received\n");
+                _receive((ng_pktsnip_t *)msg.content.ptr);
+                break;
+            case NG_NETAPI_MSG_TYPE_SND:
+            case NG_NETAPI_MSG_TYPE_GET:
+            case NG_NETAPI_MSG_TYPE_SET:
+                DEBUG("RPL: reply to unsupported recv/get/set\n");
+                reply.content.value = -ENOTSUP;
+                msg_reply(&msg, &reply);
+                break;
+            default:
+                break;
+        }
+    }
+
+    return NULL;
+}
+
+void _update_lifetime(void)
+{
+    timex_t now;
+    vtimer_now(&now);
+    ng_rpl_parent_t *parent;
+    for (uint8_t i = 0; i < NG_RPL_PARENTS_NUMOF; ++i) {
+        parent = &ng_rpl_parents[i];
+        if (parent->state != 0) {
+            if ((signed)(parent->lifetime.seconds - now.seconds) <= NG_RPL_LIFETIME_UPDATE_STEP) {
+                ng_rpl_dodag_t *dodag = parent->dodag;
+                ng_rpl_parent_remove(parent);
+                ng_rpl_parent_update(dodag, NULL);
+                continue;
+            }
+            else if (((signed)(parent->lifetime.seconds - now.seconds) <=
+                        NG_RPL_LIFETIME_UPDATE_STEP * 2)) {
```

Comment:
I don't think it's a good idea to use `NG_RPL_LIFETIME_UNIT` here. The intention is to send out a `DIS` before the parent times out. Depending on how large `NG_RPL_LIFETIME_UNIT` is, it could start too soon to send out the `DIS` packets